entities/sync/sync.go

```diff
@@ -9,104 +9,173 @@
 //  CONTACT: hello@weaviate.io
 //
 
+// Package sync provides thread-safe synchronization primitives for key-based locking.
+// It implements a memory-efficient key-based locking mechanism with automatic cleanup
+// of unused locks to prevent memory leaks. The package is particularly useful when
+// you need to synchronize access to resources identified by string keys.
 package sync
 
 import (
 	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	enterrors "github.com/weaviate/weaviate/entities/errors"
 )
 
-// KeyLocker it is a thread safe wrapper of sync.Map
-// Usage: it's used in order to lock specific key in a map
-// to synchronize concurrent access to a code block.
-//
-//	locker.Lock(id)
-//	defer locker.Unlock(id)
-type KeyLocker struct {
-	m sync.Map
+type lockEntry struct {
+	lock     sync.Mutex
+	lastUsed atomic.Int64 // Unix nano for atomic access
 }
 
-// NewKeyLocker creates Keylocker
-func NewKeyLocker() *KeyLocker {
-	return &KeyLocker{
-		m: sync.Map{},
-	}
+// KeyLocker provides thread-safe locking by key with automatic resource cleanup.
+// It maintains a bounded number of locks and automatically removes unused ones
+// based on both size limits and time-based expiration.
+//
+// Features:
+//   - Thread-safe key-based locking
+//   - Automatic cleanup of unused locks
+//   - Size-based eviction (LRU)
+//   - Time-based expiration
+//   - Memory leak prevention
+type KeyLocker struct {
+	locks   sync.Map
+	maxAge  time.Duration // maximum age of unused locks
+	count   atomic.Int32  // current count of locks
+	maxSize int32         // maximum number of locks
+	logger  logrus.FieldLogger
 }
 
-// Lock it locks a specific bucket by it's ID
-// to hold ant concurrent access to that specific item
+// NewKeyLocker creates a new KeyLocker with default settings.
+// It initializes a KeyLocker with:
+//   - Maximum size of 1000 locks
+//   - Maximum age of 24 hours for unused locks
+//   - Automatic cleanup every 5 minutes
 //
-//	do not forget calling Unlock() after locking it.
-func (s *KeyLocker) Lock(ID string) {
-	iLock := &sync.Mutex{}
-	iLocks, _ := s.m.LoadOrStore(ID, iLock)
+// The created KeyLocker automatically starts a background goroutine
+// for periodic cleanup of expired locks.
+func NewKeyLocker(logger logrus.FieldLogger) *KeyLocker {
+	kl := &KeyLocker{
+		maxSize: 1000,           // default max size
+		maxAge:  24 * time.Hour, // default max age
+		logger:  logger,
+	}
 
-	iLock = iLocks.(*sync.Mutex)
-	iLock.Lock()
-}
+	enterrors.GoWrapper(func() { // cleanup every 5 minutes
+		kl.periodicCleanup(5 * time.Minute)
+	}, kl.logger)
```

Comment:
Would it be possible to add a shutdown method on `KeyLocker` which would close this goroutine.

Fire and forget goroutines are a nightmare if you want to add goroutine leak detection checks in tests, like: https://github.com/uber-go/goleak

Reply:
this is no fire and forget go routine, it's required to live while the DB is alive, clean ups will be called on indexes shutdown though see https://github.com/weaviate/weaviate/pull/7458/files#diff-d080e5efa621477a1d974bced14c39a8912fcb7a649eb4a0fb219647728dbc3dR2132

Reply:
Yeah, but indexes can be created and deleted. And this routine will be living its life for eternity.

Additionally, any keys it is holding will be prevented from garbage collection together with this `KeyLocker` object

Reply:
yeah agreed, i think we need different approach in general